### PR TITLE
feat: appid extension in assertion challenge response

### LINKED ISSUE #106: appid extension: u2f registered keys compat
So I'm experimenting with this library. Currently the issue is with the appid extension enabled it's claiming the device registered via u2f is not familiar. Newly registered keys work with my config, but the old FIDO ones do not. This issue occurs when `navigator.credentials.get` is called in either Chrome or Firefox. Code snips:

Creating the handler:
```go
	if webauthnHandler, err = webauthn.New(&webauthn.Config{
		RPID:                  "webauthn.example.com",
		RPDisplayName:         "Webauthn",
		RPOrigin:              "https://webauthn.example.com",
		AttestationPreference: protocol.PreferIndirectAttestation,
	}); err != nil {
		panic(err)
	}
```

Credential added the users credentials:
```go
	keyID, _ := base64.StdEncoding.DecodeString(encodedKeyHandleIDHere)
	pubkey, _ := base64.StdEncoding.DecodeString(encodedPubKeyHere)

	credential := webauthn.Credential{
		ID:              keyID,
		PublicKey:       pubkey,
		AttestationType: "fido-u2f", // Also tried with this commented.
		Authenticator: webauthn.Authenticator{
			SignCount: 0,
			AAGUID:    []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		},
	}
```

Creating the assertion:
```go
	assertion, data, err := webauthnHandler.BeginLogin(user,
		webauthn.WithUserVerification(body.UserVerification),
		webauthn.WithAssertionExtensions(map[string]interface{}{"appid": "https://webauthn.example.com"}),
	)
```

### PR DESCRIPTION
**Pull request by james-d-elliott**
This change allows FinishLogin to process appid requests if the appid key in the extensions map has a value. The extensions map is expected as a root member of the JSON object sent to the endpoint.

Closes #106

Co-authored-by: Clément Michaud <clement.michaud34@gmail.com>

### CODE REVIEW

File: protocol/assertion.go
Comment (james-d-elliott) [inline protocol/assertion.go:145]:
```suggestion
	if appID == "" || attestationType != "fido-u2f" {
```

Maybe we should pass the attestation type from the loginCredential here as well for safety?

Comment (james-d-elliott) [reply protocol/assertion.go:145]: See this commit for an example, plus validation that the appid was actually set by the server by recording it in the session data:

https://github.com/authelia/webauthn/pull/7/commits/1f0ad2d885b2f8e3589a9ad7b0a542d754aefde8

**Comment by james-d-elliott** [comment]
A note regarding attestation and why I have not added any tooling for that:

`This extension does not allow FIDO-compatible credentials to be created. Thus, credentials created with WebAuthn are not backwards compatible with the FIDO JavaScript APIs.`

Clearly stated in https://www.w3.org/TR/webauthn-2/#sctn-appid-extension

**Comment by james-d-elliott** [comment]
If you decide to take this up, you can ping me to ensure it's up to date with the solution I made upstream. Otherwise you can check out what we did here: https://github.com/authelia/webauthn/commit/509e08fb364c78be30067a93d976730a8fe4a656

**Comment by MasterKale** [comment]
Howdy all, if the merge conflicts get resolved and a maintainer signs off I'll merge this.

**Comment by james-d-elliott** [comment]
> .

Yep no worries  @MasterKale, I'll resolve them this weekend and check the final implementation we went with downstream and give you a ping. I will give you some bullet points to help out with the process as well, to make the decision process regarding a few of the things I'm less confident with.

1. Should the JSON attribute for the values collected by the PublicKeyCredential.getClientExtensionResults() be assigned to the JSON key `extensions` or should we be using `clientExtensionResults`. I have noted a couple client libraries using `clientExtensionResults` but none so far using `extensions`.
2. I think we should be recording the value of appid if it was used in the session data along with the value of this option. This allows us to validate the appid properly without relying on information sent by the client. As per the [specifiction](https://w3c.github.io/webauthn/#sctn-appid-extension) the result from getClientExtensionResults is expected to have a key named appid with a boolean value, true indicating the authentication used the appid extension, false meaning it didn't. I think we should store the value of appid for this reason.

**Comment by MasterKale** [comment]
> 1. Should the JSON attribute for the values collected by the PublicKeyCredential.getClientExtensionResults() be assigned to the JSON key `extensions` or should we be using `clientExtensionResults`. I have noted a couple client libraries using `clientExtensionResults` but none so far using `extensions`.

I'd say `clientExtensionResults` since the method you use to get them is called `getClientExtensionResults()`. In my experience I think it's beneficial to not overload the term "extensions" here because I've observed that the value you get back from `getClientExtensionResults()` doesn't always reflect the [extension data you may get back in the `authData`](https://www.w3.org/TR/webauthn-2/#sctn-attestation). This makes sense in light of the spec's division between [Client Extensions](https://www.w3.org/TR/webauthn-2/#sctn-client-extension-processing) and [Authentication Extensions](https://www.w3.org/TR/webauthn-2/#sctn-authenticator-extension-processing).

> 2. I think we should be recording the value of appid if it was used in the session data along with the value of this option. This allows us to validate the appid properly without relying on information sent by the client. As per the [specifiction](https://w3c.github.io/webauthn/#sctn-appid-extension) the result from getClientExtensionResults is expected to have a key named appid with a boolean value, true indicating the authentication used the appid extension, false meaning it didn't. I think we should store the value of appid for this reason.

Speaking generally I agree with you, `appid` needs to be remembered after generating options so that you can specify it as the expected RP ID instead during registration response verification.

**Comment by james-d-elliott** [comment]
This should be ready for review @MasterKale, I believe the changes I've made are in line with both your feedback and the spec.

I have to check the renaming of the fields is not a breaking change and if it is I'll revert it and just use the tags for now and we can document that it should be renamed in v1 for clarity (speaking of it would be a good idea to introduce versioning and a policy surrounding that).

**Comment by 6543** [comment]
we at [gitea](https://github.com/go-gitea/gitea) like to get this pull in too!

**Review by aseigler — APPROVED** [review]
LGTM

**Comment by james-d-elliott** [comment]
We may also want to consider some docs regarding this as I had a question from the gitea team regarding the usage of this. Basically the attestation type of the credential has to be fido-u2f in order for the process to complete successfully. This is done due to the fact that the storage of U2F key id/handles is not in the CBOR format but in another format. We *could* theoretically rely on the values returned by the browser but I feel like it's wise since this is a security library to have as many checks that are not prohibitive as possible. I have included some small snippets from my reference/test implementation that may be of use below.

Example credential:

```go
	credential := webauthn.Credential{
		ID:              id,
		PublicKey:       pubkey,
		AttestationType: "fido-u2f",
		Authenticator: webauthn.Authenticator{
			SignCount: 0,
			AAGUID:    []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		},
	}
```

Example usage:

```go
	user := getUser()

	var opts = []webauthn.LoginOption{
		webauthn.WithUserVerification(protocol.VerificationPreferred),
	}

	extensions := make(map[string]interface{})

	for _, cred := range user.Credentials {
		if cred.AttestationType == "fido-u2f" {
			extensions["appid"] = "https://webauthn.example.com"
			break
		}
	}

	if len(extensions) != 0 {
		opts = append(opts, webauthn.WithAssertionExtensions(extensions))
	}

	var (
		assertion *protocol.CredentialAssertion
		data      *webauthn.SessionData
	)


        assertion, data, err = webauthnHandler.BeginLogin(user, opts...)
```

WebauthnHandler: 

```go
	if webauthnHandler, err = webauthn.New(&webauthn.Config{
		RPID:                  "webauthn.example.com",
		RPDisplayName:         "Webauthn",
		RPOrigin:              "https://webauthn.example.com",
		AttestationPreference: protocol.PreferNoAttestation,
	}); err != nil {
		panic(err)
	}
```

Here are the typescript interfaces I'm working off:

```typescript
interface AttestationChallengeResponse {
    id: string;
    rawId: string;
    type: string;
    clientExtensionResults: AuthenticationExtensionsClientOutputs;
    transports?: AuthenticatorTransport[];
    response: AttestationResponse;
}

interface AttestationResponse {
    attestationObject: string;
    clientDataJSON: string;
}

interface AssertionChallengeResponse {
    id: string;
    rawId: string;
    type: string;
    clientExtensionResults: AuthenticationExtensionsClientOutputs;
    response: AssertionResponse;
}

interface AssertionResponse {
    authenticatorData: string;
    clientDataJSON: string;
    signature: string;
    userHandle: string;
}
```

Here are the methods I'm using that populate them:

```typescript
    let response: AttestationChallengeResponse = {
        id: credential.id,
        rawId: bufferEncode(new Uint8Array(credential.rawId)),
        type: credential.type,
        clientExtensionResults: credential.getClientExtensionResults(),
        response: {
            attestationObject: bufferEncode(new Uint8Array(attestationResponse.attestationObject)),
            clientDataJSON: bufferEncode(new Uint8Array(attestationResponse.clientDataJSON)),
        },
    };
    
    let response: AssertionChallengeResponse = {
        id: credential.id,
        rawId: bufferEncode(new Uint8Array(credential.rawId)),
        type: credential.type,
        clientExtensionResults: credential.getClientExtensionResults(),
        response: {
            authenticatorData: bufferEncode(new Uint8Array(assertionResponse.authenticatorData)),
            clientDataJSON: bufferEncode(new Uint8Array(assertionResponse.clientDataJSON)),
            signature: bufferEncode(new Uint8Array(assertionResponse.signature)),
            userHandle: userHandle,
        },
    };
```

**Comment by 6543** [comment]
gitea pull based on this branch got merged

**Comment by james-d-elliott** [comment]
Yep, I saw, I will make sure to make a PR downstream or notify you if this PR changes in a meaningful way.

File: protocol/authenticator.go
Comment (zachhuff386) [inline protocol/authenticator.go:226]: This should probably be `return ErrVerification.WithInfo(fmt.Sprintf("RP Hash mismatch. Expected %v but Received %v and %v\n", a.RPIDHash, rpIdHash, appIDHash))` to log the U2F app ID hash and use `%v` to prevent putting binary data in an error message.

Comment (james-d-elliott) [reply protocol/authenticator.go:226]: Good spot, though maybe %x  would be better since %v will just print the uint8 values?

**Comment by james-d-elliott** [comment]
@MasterKale Any chance at another look at this? Or is there something you're waiting on?

**Comment by MasterKale** [comment]
@aseigler gave his stamp of approval already, I suppose I'd like to see @zachhuff386's review get resolved before I mash the Merge button. Just to reiterate I'm not capable of reviewing this repo, I just have the permission it takes to merge PR's. I'm relying on established Golang-capable maintainers to indicate to me what's good to merge or not.

**Comment by james-d-elliott** [comment]
> @aseigler gave his stamp of approval already, I suppose I'd like to see @zachhuff386's review get resolved before I mash the Merge button. Just to reiterate I'm not capable of reviewing this repo, I just have the permission it takes to merge PR's. I'm relying on established Golang-capable maintainers to indicate to me what's good to merge or not.

Technically it's already resolved if he accepts my proposal of `%x` instead of `%v`. The difference is %v outputs `[0, 1, 213, 881, ....]` for bytes I believe and %x will output the hex values.

File: protocol/assertion.go
Comment (zachhuff386) [inline protocol/assertion.go:149]: I am using the code to support U2F keys and everything is working. But to be safe the error from `webauthncose.VerifySignature` shouldn't be ignored. There isn't any outcome that causes valid to be `true` with an error but both should be checked with `!valid || err != nil`. The errors from `webauthncose.ParsePublicKey` and `webauthncose.ParseFIDOPublicKey` are also unhandled.

```go
	if appID == "" {
		key, err = webauthncose.ParsePublicKey(credentialBytes)
	} else {
		key, err = webauthncose.ParseFIDOPublicKey(credentialBytes)
	}
	if err != nil {
		return err
	}

	valid, err := webauthncose.VerifySignature(key, sigData, p.Response.Signature)
	if !valid || err != nil {
		return ErrAssertionSignature.WithDetails(fmt.Sprintf("Error validating the assertion signature: %+v\n", err))
	}
```

Also unrelated to this pull request the challenge compare here https://github.com/duo-labs/webauthn/blob/master/protocol/client.go#L80 could be done with constant time compare to prevent any potential timing attacks. Changing `if 0 != strings.Compare(storedChallenge, challenge) {` to `if subtle.ConstantTimeCompare([]byte(storedChallenge), []byte(challenge)) != 1 {`.

File: protocol/authenticator.go
Comment (james-d-elliott) [reply protocol/authenticator.go]: Did you have a preference on this? @zachhuff386

File: protocol/assertion.go
Comment (james-d-elliott) [reply protocol/assertion.go:149]: We can definitely do that, I had in mind to ask about this actually since it seemed odd not to check errors.

Comment (james-d-elliott) [reply protocol/assertion.go:149]: Good spot on the comparison. This has been added and the other changes have been pushed. If you could check if you think the changes are sufficient (the last two commits) that'd be good.

File: protocol/client.go
Comment (zachhuff386) [inline protocol/client.go:81]: This should be `!= 1` not `== 1`, it returns 1 on a match.

Comment (james-d-elliott) [reply protocol/client.go]: Ah yep, glad you spotted this, wish I did! Completely forgot there was a significant API difference between the two methods.

**Comment by MasterKale** [comment]
@james-d-elliott Is this good to go?

**Pull request merged**
